Protect kinematic formulation against scaling effects

lindsayad opened the pull request:
I'll be adding a test shortly. @bwspenc in the process of doing this work I realized that RANFS and the kinematic formulation are indeed different whenever the normal vector is not perfectly aligned along a Cartesian axis; if the latter condition is not met, then kinematic is really indeed a penalty formulation. I will demonstrate that in my test.

Also, I think that there should be some similar protections for the tangential_penalty method because it seems to be a hybrid of kinematic and true penalty. But that information is kind of confounded in the `_contact_force` so I didn't touch it here.

to-do list:
- [x] RANFS frictionless
- [x] kinematic frictionless
- [x] tangential penalty frictional (which uses kinematic for the zero-penetration constraint)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/447285/) on f97ef5e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14546/site)

This comment will be updated on new commits.

tophmatthews commented:
> I realized that RANFS and the kinematic formulation are indeed different whenever the normal vector is not perfectly aligned along a Cartesian axis

Is the expectation that RANFS will replace the kinematic formulation? We should probably add a suite of tests, much like I did for mortar+tensor mechanics.

tophmatthews commented:
> Is the expectation that RANFS will replace the kinematic formulation? We should probably add a suite of tests, much like I did for mortar+tensor mechanics.

???

lindsayad commented:
> Is the expectation that RANFS will replace the kinematic formulation? We should probably add a suite of tests, much like I did for mortar+tensor mechanics.

I strongly believe it is the best node-face-based algorithm that we have. Note that I have not yet done anything with frictional. RANFS is currently only frictionless.

tophmatthews commented:
> I strongly believe it is the best node-face-based algorithm that we have.

I like the confidence. I'll try and get some tests going similar to what I did for mortar method to tease out any remaining problems.



> RANFS is currently only frictionless.

What is the timeline for RANFS for a) glued b)frictionless c) pull-away?

lindsayad commented:
Sadly, the node-face RANFS algorithm is susceptible to ping-ponging, just as kinematic is, as shown in https://github.com/idaholab/moose/pull/14546/commits/76fe09f153179b042ae94046684ae582f32925b8. Mortar appears to be more robust since it does not show ping-ponging at any time step for that particular test input while RANFS and kinematic node-face do.

So despite my optimism in https://github.com/idaholab/moose/pull/14241#issuecomment-549105944, node-face RANFS is not the magic bullet. Your ping-ponging test in #14241 also shows that while mortar may-be more robust than node-face with respect to ping-ponging, it is still not immune. In fact [my favorite mortar contact paper](https://onlinelibrary.wiley.com/doi/pdf/10.1002/gamm.201410004) shows an example of ping-ponging mortar. They appear to have been able to fix that case by using a Petrov-Galerkin approach. So that's my newest golden idea 😄

tophmatthews commented:
Dude...you're killing me with this whiplash... :)

moosebuild commented:
Job [Boost enabled, Pthreads only](https://civet.inl.gov/job/449862/) on 76fe09f : invalidated by @lindsayad

Some weird bs

permcody commented:
Topher or Ben, could one of you perform the review on this PR?

Thanks!

bwspenc commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -51,6 +51,10 @@ validParams<ContactAction>()
+  params.addParam<Real>("tangential_penalty",
+                        1e8,
+                        "The tangential penalty to apply for frictional contact.  This can vary "
+                        "depending on the stiffness of your materials");
```
Getting these parameters right is a can of worms. A few issues here:
1. We should probably disallow the use of `tangential_penalty` for frictionless and glued contact because it doesn't do anything in those case, so it's misleading to provide it as an option.
2. Default values for either `tangential_penalty` or `penalty` are problematic because they are problem-dependent. Maybe we should make `penalty` required, and have an optional multiplier on that for the tangential component that defaults to 1. Auto scaling makes this even messier.
3. The descriptions for 'penalty' and 'tangential_penalty' aren't very helpful. We can't explain all the intricacies of how to use them in different scenarios in one line, but we need to make sure that is covered in the .md document, and we should do better here too.

lindsayad replied:
>     2\. Maybe we should make `penalty` required, and have an optional multiplier on that for the tangential component that defaults to 1.

I think we should be moving away from use of `penalty` in input files, not making it required (obviously with the exception of when we are actually using a `ContactFormulation::PENALTY`). If the user is using automatic scaling, as of this PR we error out if the user sets a non-unity penalty for kinematic formulations. Requiring them to set a parameter that we require to be 1 in a lot of cases seems silly. Moreover, mortar contact and RANFS contact do not use penalty parameters at all. I would suggest that if we were to change anything for the `penalty` parameter it would be to make its default value 1. I think `tangential_penalty` should be its own parameter because in the future I would love to see an input block for a `tangential_penalty` formulation that has no `penalty` parameter at all since the user shouldn't have to set it if they are using automatic scaling.

lindsayad replied:
>     3\. The descriptions for 'penalty' and 'tangential_penalty' aren't very helpful. We can't explain all the intricacies of how to use them in different scenarios in one line, but we need to make sure that is covered in the .md document, and we should do better here too.

`tangential_penalty` isn't that complicated is it? It always corresponds to a true penalty concept right? Or is there something weird for Augmented Lagrange? I vaguely remember you saying something related to that. I don't know anything about `tangential_penalty` other than that it shouldn't be equal to the `penalty` for a `TANGENTIAL_PENALTY` formulation

lindsayad replied:
In the past, have you typically used the same order of magnitude `penalty` for both `KINEMATIC` and `PENALTY`?

bwspenc replied:
Yes, in the past they have been of a similar order of magnitude.  I'm just saying here that since these parameters are used different ways for the different formulations, it's more than you can explain in a short parameter description.  Also keep in mind that these behave differently when you use the normalize_penalty option, which also needs to be explained.  That sentence about it varying based on the stiffness of the materials isn't terribly descriptive in my opinion.

lindsayad replied:
> That sentence about it varying based on the stiffness of the materials isn't terribly descriptive in my opinion.

But I feel like if you're going to give a one sentence description, that sentence is not a bad one. The only way I would improve it given a concision constraint is to say this would be roughly on the order of magnitude of the stiffness.

bwspenc commented on modules/contact/src/constraints/MechanicalContactConstraint.C:
```diff
@@ -1288,7 +1328,11 @@ MechanicalContactConstraint::computeQpOffDiagJacobian(Moose::ConstraintJacobianT
+  if (_formulation == ContactFormulation::KINEMATIC ||
+      _formulation == ContactFormulation::TANGENTIAL_PENALTY)
+    penalty_temp /= _var.scalingFactor();
+  const Real penalty = penalty_temp;
```
Same change I suggested for computeQpJacobian here.

bwspenc commented on modules/contact/test/tests/bouncing-block-contact/ping-ponging/kinematic-ping-pong.i:
```diff
@@ -0,0 +1,109 @@
+
+[Kernels]
+  [./disp_x]
+    type = MatDiffusion
```
I'm fine with this as a temporary solution, but can't we just make the `contact` module depend on `tensor_mechanics` so that these can be real problems? We could move all of our contact test cases from the `combined` module if we did that. Thoughts on this, @permcody? We're already doing that with the `xfem` module for similar reasons. That would also allow us to get a meaningful code coverage report from the `contact` module.

tophmatthews replied:
We could move all of the combined/test/tests/mortar_tm tests here, and the subsequent tests I'll add shortly for frictional, RANFS, kinematic, etc.

While you're at it, should the thermal contact actions be moved to contact, and have contact depend on heat_conduction?

lindsayad replied:
> That would also allow us to get a meaningful code coverage report from the `contact` module.

This would provide added coverage for the old contact formulations/models. Mortar and RANFS objects are already covered in the current `contact` module. I'm supportive of creating the `tensor_mechanics` dependency...but I don't want to do that in this PR.

bwspenc replied:
I wouldn't advocate for mixing mechanical and thermal contact in the modules.  That would make it so an app that brings in the 'contact' module would also get heat_conduction, which they may not want.  I can't imagine an app bringing in 'contact' and not wanting 'tensor_mechanics', though.

bwspenc replied:
I could see the argument for just moving all of the contact code into 'tensor_mechanics', though.  Then it would be consistent with how 'heat_conduction' is structured.

bwspenc replied:
I just merged a PR in today that makes contact depend on tensor_mechanics.  It would make me feel so much better if these all used tensor_mechanics.

bwspenc reviewed (changes requested):
A few nit-picky things. I think the only major thing to resolve is the issues around the `penalty` and `tangential_penalty` parameters.

bwspenc commented on modules/contact/doc/content/modules/contact/index.md:
```diff
@@ -270,3 +270,25 @@ The recommended PETSc options for use with `NodeFaceConstraint` based contact ar
+| Tangential Penalty no scaling AMG | No convergence | No convergence |
+| Tangential Penalty auto-scaling AMG | 53 | 208 |
+| Tangential Penalty no scaling LU | 60 | 224 |
+| Tangential Penalty auto-scaling LU | 41 | 116 |
```
Also, these results are going to be out of date the second this gets merged.  It's useful information, but I'm a little hesitant to include stuff like this in the documentation.

tophmatthews replied:
I think the trends are still valid since these are "arbitrary units"

lindsayad replied:
> Also, these results are going to be out of date the second this gets merged

What do you mean by this? The numbers may change very slightly with changes in PETSc version, but the changes should be small. I feel like quantitative comparisons like this are very valuable for showing the user that certain things are important, like scaling. I think these results are much more poignant than simply saying "scaling improves performance".

bwspenc replied:
I agree that it's better to include real numbers.  Maybe you could provide a date or something in the text to indicate when those numbers were recorded.  Based on my experience with these kinds of things, they will not change just slightly over time.  There are lots of moving parts in the code that affect these things.

lindsayad commented on modules/contact/test/tests/kinematic-and-scaling/tests:
```diff
@@ -0,0 +1,26 @@
+    type = Exodiff
+    input = bouncing-block-kinematic.i
+    exodiff = bouncing-block-kinematic_exo.e
+    requirement = 'The system shall be able to apply automatic scaling in conjunection with kinematic contact constraint enforcement and show zero penetration with a unity penalty factor'
```
It's close to axis-aligned. I will change this to say "close to no penetration"

permcody commented:
@bwspenc - Looks like all of your comments have been addressed or rebutted. Any further comments or is this good to go?

bwspenc commented:
This all looks good to me except for the issue with the parameters.

lindsayad commented:
K I think this PR is mostly a crutch that doesn't really address what seems to be a major issue with PJFNK and scaling when:

- at least one physics has dramatic differences in magnitude between Jacobian and residual values
- There are at least two physics

My test cases in this PR are single physics and don't really test this case.

Here is a slightly modified version of @acasagran's input that has temperature and displacement variables:

<details>
  <summary>Input</summary>

```
[GlobalParams]
  displacements = 'disp_x disp_y'
  order = FIRST
  family = LAGRANGE
  temperature = temp
[]

[Problem]
  coord_type = RZ
  type = ReferenceResidualProblem
  extra_tag_vectors = 'ref'
  reference_vector = 'ref'
  group_variables = 'disp_x disp_y'
[]

[Mesh]
  file = three_pellet_2D_smeared.e
  patch_size = 5
[]

[Variables]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
  [./temp]
    initial_condition = 298
  [../]
[]

[AuxVariables]
  [./gap_conductance]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[Functions]
  [./coolant_press_ramp]
    type = PiecewiseLinear
    x = '0 1.0e5 5.1e5'
    y = '0  1.0  1.0'
  [../]
  [./coolant_temp_ramp]
    type = PiecewiseLinear
    x = '0 1.0e5 5.0e5 5.1e5'
    y = '300.0  680.0  680.0 300.0'
  [../]
  [./volumetric_heat]
    type = PiecewiseLinear
    x = '0 1.0e5 2.0e5 3.0e5 4.0e5 5.0e5 5.1e5'
    y = '0  5.0e8 1.0e9 2.0e9 5.0e9 0  0'
  [../]
[]

[Modules]
  [./TensorMechanics]
    [./Master]
      [./fuel]
        strain = FINITE
        block = '3'
        add_variables = true
        incremental = true
        eigenstrain_names = fuel_eigenstrain
        generate_output = 'stress_xx stress_yy vonmises_stress elastic_strain_yy'
        extra_vector_tags = 'ref'
      [../]
      [./clad]
        strain = FINITE
        block = '1'
        add_variables = true
        incremental = true
        generate_output = 'stress_xx stress_yy vonmises_stress elastic_strain_yy creep_strain_yy'
        extra_vector_tags = 'ref'
      [../]
    [../]
  [../]
[]

[Kernels]
  [./heat]
    type = HeatConduction
    variable = temp
    extra_vector_tags = 'ref'
  [../]
  [./heat_ie]
    type = HeatConductionTimeDerivative
    variable = temp
    extra_vector_tags = 'ref'
  [../]
  [./heat_source_fuel]
    type = HeatSource
    variable = temp
    block = '3'
    function = volumetric_heat
    extra_vector_tags = 'ref'
  [../]
[]

[AuxKernels]
  [./gap_cond]
    type = MaterialRealAux
    property = gap_conductance
    variable = gap_conductance
    boundary = 10
    execute_on = linear
  [../]
[]

[Contact]
  [./pellet_clad_mechanical]
    master = 5
    slave = 10
    system = Constraint
    penalty = 1e7
    model = frictionless
    formulation = kinematic
    normal_smoothing_distance = 0.1
  [../]
[]

[ThermalContact]
  [./pellet_clad_thermal]
    type = GapHeatTransfer
    variable = temp
    master = 5
    slave = 10
    gap_conductivity = 1.0
    quadrature = true
  [../]
[]

[BCs]
  [./no_x_all]
    type = PresetBC
    variable = disp_x
    boundary = 12
    value = 0.0
  [../]
# pin entire clad bottom in y
  [./no_y_clad_bottom]
    type = PresetBC
    variable = disp_y
    boundary = 1
    value = 0.0
  [../]
# pin fuel bottom in y
  [./no_y_fuel_bottom]
    type = PresetBC
    variable = disp_y
    boundary = '1020'
    value = 0.0
  [../]
# wedge bcs
  [./clad_temp]
    type = FunctionPresetBC
    variable = temp
    boundary = '1 2 3'
    function = coolant_temp_ramp
  [../]
  [./Pressure]
    # apply coolant pressure on clad outer walls
    [./coolantPressure]
      boundary = '1 2 3'
      factor = 15.5e6
      function = coolant_press_ramp
    [../]
  [../]
[]

[Materials]
  [./fuel_thermal]
    type = HeatConductionMaterial
    block = '3'
    thermal_conductivity = 20
    specific_heat = 500.0
  [../]
  [./fuel_elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    block = '3'
    youngs_modulus = 2e11
    poissons_ratio = 0.345
  [../]
  [./fuel_thermal_strain]
    type = ComputeThermalExpansionEigenstrain
    block = '3'
    stress_free_temperature = 298
    thermal_expansion_coeff = 50.0e-6
    eigenstrain_name = fuel_eigenstrain
  [../]
  [./fuel_stress]
    type = ComputeFiniteStrainElasticStress
    block = '3'
  [../]
  [./radial_return_stress]
    type = ComputeMultipleInelasticStress
    block = 1
    inelastic_models = 'power_law_creep'
    tangent_operator = elastic
  [../]
  [./power_law_creep]
    type = PowerLawCreepStressUpdate
    block = 1
    coefficient = 1.0e-15
    n_exponent = 1
    activation_energy = 0
  [../]
  [./clad_elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    block = 1
    youngs_modulus = 7.5e10
    poissons_ratio = 0.3
  [../]
  [./clad_thermal]
    type = HeatConductionMaterial
    block = 1
    thermal_conductivity = 16.0
    specific_heat = 330.0
  [../]
  [./clad_density]
    type = Density
    block = 1
    density = 6551.0
  [../]
  [./fuel_density]
    type = Density
    block = '3'
    density = 10000.0
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'

  petsc_options = '-ksp_monitor_true_residual -pc_svd_monitor'
  # petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  # petsc_options_value = 'lu       superlu_dist'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'svd'

  line_search = 'none'

  l_max_its = 50
  l_tol = 1e-3
  nl_max_its = 40
  nl_rel_tol = 1e-4

  dtmax = 1.0e5
  dtmin = 1.0
  end_time = 2.0e5

  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 1e3
    optimal_iterations = 20
    iteration_window = 4
    time_t ='0   5.0e4 1.0e5 5.1e5'
    time_dt ='2e3  1.2e4  7.5e3   7.5e3'
  [../]

  [./Quadrature]
     order = fifth
     side_order = seventh
  [../]

  verbose = true
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Postprocessors]
  [./ave_temp_interior]
    type = SideAverageValue
    boundary = 9
    variable = temp
    execute_on = 'initial linear'
  [../]
  [./clad_inner_vol]
   type = InternalVolume
    boundary = 7
    execute_on = 'initial timestep_end'
  [../]
  [./pellet_volume]
    type = InternalVolume
    boundary = 8
    execute_on = 'initial timestep_end'
  [../]
  [./gas_volume]
    type = InternalVolume
    boundary = 9
    execute_on = 'initial linear'
  [../]
  [./avg_clad_temp]
    type = SideAverageValue
    boundary = 7 #bws was 9
    variable = temp
    execute_on = 'initial timestep_end'
  [../]
  [./flux_from_clad]           # area integrated heat flux from the cladding
    type = SideFluxIntegral
    variable = temp
    boundary = 5
    diffusivity = thermal_conductivity
    execute_on = linear
  [../]
  [./flux_from_fuel]          # area integrated heat flux from the fuel
    type = SideFluxIntegral
    variable = temp
    boundary = 10
    diffusivity = thermal_conductivity
    execute_on = linear
  [../]
  [./dt]
    type = TimestepSize
  [../]
  [./residual]
    type = Residual
  [../]
  [./nl_its]
    type = NumNonlinearIterations
  [../]
  [./lin_its]
    type = NumLinearIterations
  [../]
  [cum_nl]
    type = CumulativeValuePostprocessor
    postprocessor = nl_its
  []
  [cum_lin]
    type = CumulativeValuePostprocessor
    postprocessor = lin_its
  []
  [./clad_elongation]
    type = NodalMaxValue
    variable = disp_y
    boundary = 2
  [../]
[]

[Outputs]
  csv = true
  interval = 1
  exodus = true
  color = false
  print_linear_residuals = false
  [./console]
    type = Console
    perf_log = true
    solve_log = true
  [../]
[]
```
</details>

Here are solve log files:

[svd-no-scaling.log](https://github.com/idaholab/moose/files/4056930/svd-no-scaling.log)
[svd-with-fix-auto-scaling.log](https://github.com/idaholab/moose/files/4056931/svd-with-fix-auto-scaling.log)
[svd-without-fix-auto-scaling.log](https://github.com/idaholab/moose/files/4056932/svd-without-fix-auto-scaling.log)

Even with this PR the no scaling case is more performant than the auto scaling case with the former taking 36 non-inear iterations while the latter takes 85 cumulative non-linear iterations. Without this PR, the input file does not run to completion when using automatic scaling. This PR allows the input to run to completion with auto-scaling because it essentially scales up the constraint residuals by 1e3-1e5, which I hypothesize allows the displacements to compete with the temperature residual in order to determine a better Newton step for the entire system; in general the displacement residual almost always starts near 1e-8 while the temperature residual is near unity. Note that contact comes into play between the second and third nonlinear iterations of Time Step 5; however, convergence is degraded for the auto-scaling case well before that! Thus I think we can probably drum up a test case not involving contact which would allow exploration of this scaling issue.

tophmatthews commented:
The only thing missing in order to newton is mechanical contact no?  Thermal contact here will likely be replaced with mortar?

lindsayad commented:
> The only thing missing in order to newton is mechanical contact no? Thermal contact here will likely be replaced with mortar?

I believe that's true

lindsayad commented:
Unfortunately, it seems that even for single physics with Newton the residual scaling is important. Take this input based on @acasagran's new BISON training which is mechanics only (no heat conduction, no contact):

<details>
<summary>mechanics only</summary>

```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  volumetric_locking_correction = true
[]

[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 5
  ny = 1
  nz = 1
  xmin = 0.0
  xmax = 4.0
  ymin = 0.0
  ymax = 0.5
  zmin = 0.0
  zmax = 1.0
  elem_type = HEX8
[]

[Modules/TensorMechanics/Master]
  [./all]
    strain = FINITE
    add_variables = true
    generate_output = 'stress_xx stress_xy stress_yy strain_xx strain_xy strain_yy'
  [../]
[]

[Functions]
  [./loading_func]
    type = PiecewiseLinear
    x = '0.  5.'
    y = '0. 50.0'
  [../]
[]

[BCs]
  [./free_end_moment]
    type = Pressure
    variable = disp_y
    component = 0
    boundary = right
    factor = 1
    function = loading_func
  [../]
  [./FixedCenterLineX]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0.0
  [../]
  [./FixedCenterLineY]
    type = DirichletBC
    variable = disp_y
    boundary = left
    value = 0.0
  [../]
  [./FixedCenterLineZ]
    type = DirichletBC
    variable = disp_z
    boundary = left
    value = 0.0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1e6
    poissons_ratio = 0.3
  [../]
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]
[]

[Preconditioning]
  [./pc]
    type = SMP
    full = True
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu     superlu_dist'

  l_max_its = 20
  l_tol = 1e-2

  nl_max_its = 50
  # nl_rel_tol = 1e-8
  # nl_abs_tol = 1e-4

  start_time = 0.0
  dt = 1
  end_time = 5
[]

[Postprocessors]
  [./num_lin_it]
    type = NumLinearIterations
  [../]
  [./num_nonlin_it]
    type = NumNonlinearIterations
  [../]
  [./tot_lin_it]
    type = CumulativeValuePostprocessor
    postprocessor = num_lin_it
  [../]
  [./tot_nonlin_it]
    type = CumulativeValuePostprocessor
    postprocessor = num_nonlin_it
  [../]
  [./alive_time]
    type = PerfGraphData
    section_name = Root
    data_type = TOTAL
  [../]
  [./max_beam_deflection]
    type = NodalMaxValue
    variable = disp_y
    boundary = 'right'
  [../]
[]

[Outputs]
  exodus = true
  csv = true
[]
```
</details>
Without automatic scaling, the condition number using `-pc_type svd -pc_svd_monitor` is 1e7. The total number of nl iterations is 44. With automatic scaling, the condition number improves to 1e4. Good job automatic scaling! However, the number of nonlinear iterations actually increases to 49. I believe this is because with automatic scaling the initial residual is of order 1e-6 and in order to achieve convergence we start hitting around machine epsilon and we lose the second order convergence of Newton. So maybe I should try this with #14422...

tophmatthews commented:
Or reference residual?

lindsayad commented:
> Or reference residual?

This is single physics and it's not near a steady-state so `ReferenceResidualProblem` suffers from the same issue

lindsayad commented:
Ok, somehow I forgot that while there wasn't a bug in our concept of applying scaling in contact, there *was* indeed a bug in the jacobians, e.g. while we applied the scaling factor in the residual, we didn't do it for the Jacobian. So with that bug fixed, things look pretty good:

| Scheme | Cumulative nonlinear iterations | Cumulative linear iterations |  Condition number | Scaling Factor |
| --- | --- | --- | --- | --- |
| SMP PJFNK RANFS no scaling AMG | 61 | 310 | 6e3 | 1 |
| SMP PJFNK RANFS residual auto-scaling AMG | 55 | 226 | 1e1 | 1.7e-2 |
| SMP PJFNK RANFS Jacobian auto-scaling AMG | 56 | 228 | 2e1 | 3.7e-4 |
| SMP PJFNK Kinematic no scaling AMG | 53 | 207 | 6e3 | 1 |
| SMP PJFNK Kinematic residual auto-scaling AMG | 53 | 207 | 1e2 | 1.7e-2 |
| SMP PJFNK Kinematic Jacobian auto-scaling AMG | 53 | 207 | 1e1 | 3.7e-4 |
| SMP PJFNK Kinematic no scaling LU | 41 | 115 | 6e3 | 1 | 6e3 | 1
| SMP PJFNK Kinematic residual auto-scaling LU | 41 | 115 | 1e2 | 1.7e-2 |
| SMP PJFNK Kinematic Jacobian auto-scaling LU | 41 | 115 | 1e1 | 3.7e-4 |
| SMP PJFNK Tangential Penalty no scaling AMG | 61 | 293 | 6e3 | 1 |
| SMP PJFNK Tangential Penalty residual auto-scaling AMG | 61 | 293 | 1e2 | 1.7e-2 |
| SMP PJFNK Tangential Penalty Jacobian auto-scaling AMG | 61 | 293 | 1e1 | 3.7e-4 |
| SMP PJFNK Tangential Penalty no scaling LU | 50 | 198 | 6e3 | 1 |
| SMP PJFNK Tangential Penalty residual auto-scaling LU | 50 | 198 | 1e2 | 1.7e-2 |
| SMP PJFNK Tangential Penalty Jacobian auto-scaling LU | 50 | 198 | 1e1 | 3.7e-4 |
| FD PJFNK RANFS residual auto-scaling LU | 41 | 48 | 1e1 | 1.7 e-2 |
| FD PJFNK Kinematic residual auto-scaling LU | 56 | 80 | 1e2 | 1.7e-2 |

## Important summary notes:

- With the Jacobian fix, the solve efficiency of kinematic and tangential penalty formulations is independent of scaling (within the window of this problem)
- Because RANFS constraint residuals/Jacobians are of gap magnitude, the RANFS solve really does perform better when the internal physics is scaled
- There are some bugs with the RANFS Jacobians because with user provided Jacobian it takes more non-linears and linears than kinematic, but with a FD Jacobian it takes significantly less non-linears and linears
- I feel much better about scaling and contact now...we can keep the old meaning of the kinematic penalty parameter and scaling doesn't make the solve worse (or better within the context of this problem)

lindsayad commented:
@aeslaughter The markdown in https://github.com/idaholab/moose/pull/14546#issuecomment-578355109 renders fine on GitHub and in MacDown. Do you have any suggestion for how to fix the doc error [here](https://civet.inl.gov/job/456605/)

aeslaughter commented:
@lindsayad I am not sure, but will take a look and push a fix.

aeslaughter commented on modules/contact/doc/content/modules/contact/index.md:
```diff
@@ -271,24 +271,37 @@ The recommended PETSc options for use with `NodeFaceConstraint` based contact ar
+running with two processes.
+
+| Scheme | Cumulative nonlinear iterations | Cumulative linear iterations |  Condition number | Scaling Factor |
+| - | - | - | - | - |
```
@lindsayad I bet this fixed it, I can get it to fail.

lindsayad replied:
I actually had a row that had two many columns...sorry I should have notified you that I fixed it

lindsayad commented:
Ok @bwspenc this is ready for review again

bwspenc commented:
@lindsayad When we talked last week, it sounded like you concluded that this didn't help, and that it was better to not do this. Did something change since then?

bwspenc commented:
Oh, now I see your detailed note about the Jacobian fixes. I'll look at this today.

lindsayad commented:
Today can have nebulous meaning...

lindsayad commented:
Anyone else want to review this because it seems like @bwspenc isn't going to :cry:

bwspenc commented on modules/contact/test/tests/tan-pen-and-scaling/tests:
```diff
@@ -0,0 +1,22 @@
+    input = bouncing-block-tan-pen.i
+    exodiff = no-scaling.e
+    requirement = 'The system shall show penetration when automatic scaling is not used with tangential penalty contact constraint enforcement and when the element Jacobian diagonals are significantly larger than the (normal) penalty factor'
+    cli_args = "Executioner/automatic_scaling=false Outputs/file_base=no-scaling Postprocessors/active='' -pc_type lu -pc_factor_shift_type NONZERO"
```
Why do you need `-pc_factor_shift_type NONZERO` on problems like this?

lindsayad replied:
> Why do you need `-pc_factor_shift_type NONZERO` on problems like this?

When things weren't working well I needed to use `lu` in order for this to converge in parallel. But now that things are hunky dorey, I've removed it :smile:

bwspenc replied:
Good to hear!

bwspenc commented on modules/contact/doc/content/modules/contact/index.md:
```diff
@@ -270,3 +270,38 @@ The recommended PETSc options for use with `NodeFaceConstraint` based contact ar
+
+| Scheme | Cumulative nonlinear iterations | Cumulative linear iterations |  Condition number | Scaling Factor |
+| - | - | - | - | - |
+| SMP PJFNK RANFS no scaling AMG | 61 | 310 | 6e3 | 1 |
```
Shouldn't you just say 'N/A' for the scaling factor for the RANFS cases?

lindsayad replied:
No because `Scaling Factor` refers to the number used for variable scaling. I'll do this change: `Scaling Factor` -> `Variable Scaling Factor`

bwspenc replied:
OK, yeah -- when I first saw this, I thought you were talking about a penalty factor.

bwspenc commented on modules/contact/doc/content/modules/contact/index.md:
```diff
@@ -270,3 +270,38 @@ The recommended PETSc options for use with `NodeFaceConstraint` based contact ar
+[tangential penalty](/bouncing-block-tan-pen.i),
+running with two processes.
+
+| Scheme | Cumulative nonlinear iterations | Cumulative linear iterations |  Condition number | Scaling Factor |
```
Is 'Scaling Factor' the number that auto scaling computed? Was the same factor applied to all displacement variables?

lindsayad replied:
Yes and yes. This PR adds the ability to group variable together for determination of scaling factors. All input files using automatic scaling have the displacement variables grouped.

bwspenc commented on modules/contact/src/constraints/RANFSNormalMechanicalContact.C:
```diff
@@ -36,6 +36,10 @@ validParams<RANFSNormalMechanicalContact>()
+  params.addParam<bool>("ping_pong_protection",
+                        false,
+                        "Whether to protect against ping-ponging by tying the slave node to the "
+                        "edge between the involved master faces");
```
What kind of protective equipment do ping pong players wear? Wrist guards? Knee pads? Helmets? Protective eyewear?

lindsayad replied:
Haha, I've updated the doc string to be a little more explicit about what ping-ponging is. I guess I could call it oscillation protection...

bwspenc replied:
Nah, I think I'd leave it how you have it.

bwspenc commented on modules/contact/test/tests/tan-pen-and-scaling/bouncing-block-tan-pen.i:
```diff
@@ -0,0 +1,113 @@
+  line_search = 'none'
+  automatic_scaling = true
+  verbose = true
+  scaling_group_variables = 'disp_x disp_y'
```
How important is it to group the variables for scaling?  Is this something we would want to do even for non-contact problems?

lindsayad replied:
I think so. I have this input file from Al:

<details>

<summary>Al input</summary>

```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  volumetric_locking_correction = true
  # scaling = 1e0
[]

[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 5
  ny = 1
  nz = 1
  xmin = 0.0
  xmax = 4.0
  ymin = 0.0
  ymax = 0.5
  zmin = 0.0
  zmax = 1.0
  elem_type = HEX8
  uniform_refine = 1
[]

[Variables]
  [disp_x]
    # scaling = 3.25447e-06
  []
  [disp_y]
    # scaling = 2.29038e-06
  []
  [disp_z]
    # scaling = 3.60462e-06
  []
[]

[Modules/TensorMechanics/Master]
  [./all]
    strain = FINITE
    generate_output = 'stress_xx stress_xy stress_yy strain_xx strain_xy strain_yy'
  [../]
[]

[Functions]
  [./loading_func]
    type = PiecewiseLinear
    x = '0.  5.'
    y = '0. 50.0'
  [../]
[]

[BCs]
  [./free_end_moment]
    type = Pressure
    variable = disp_y
    component = 0
    boundary = right
    factor = 1
    function = loading_func
  [../]
  [./FixedCenterLineX]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0.0
  [../]
  [./FixedCenterLineY]
    type = DirichletBC
    variable = disp_y
    boundary = left
    value = 0.0
  [../]
  [./FixedCenterLineZ]
    type = DirichletBC
    variable = disp_z
    boundary = left
    value = 0.0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1e6
    poissons_ratio = 0.3
  [../]
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]
[]

[Preconditioning]
  [./pc]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  petsc_options_iname = '-ksp_gmres_restart'
  petsc_options_value = '200'

  l_max_its = 200

  nl_max_its = 20

  start_time = 0.0
  dt = 1
  dtmin = 1
  end_time = 5
  num_steps = 1
[]

[Postprocessors]
  [./num_lin_it]
    type = NumLinearIterations
  [../]
  [./num_nonlin_it]
    type = NumNonlinearIterations
  [../]
  [./tot_lin_it]
    type = CumulativeValuePostprocessor
    postprocessor = num_lin_it
  [../]
  [./tot_nonlin_it]
    type = CumulativeValuePostprocessor
    postprocessor = num_nonlin_it
  [../]
  [./alive_time]
    type = PerfGraphData
    section_name = Root
    data_type = TOTAL
  [../]
  [./max_beam_deflection]
    type = NodalMaxValue
    variable = disp_y
    boundary = 'right'
  [../]
[]

[Outputs]
  exodus = true
  csv = true
  print_linear_residuals = true
[]
```

</details>

where if I do normal automatic scaling the residual is 2.718520e-12 after 11 nonlinear iterations. (initial residual of 2.147227e-06). The automatic scaling factors are:
```
Automatic scaling factors:
  disp_x: 1.62724e-06
  disp_y: 1.14519e-06
  disp_z: 1.80231e-06
```
If I group them together, the residual after 11 nonlinear iterations is 8.083608e-15 with a single scaling factor of `1.14519e-06`.

I think that for the same physics it might be important to apply the same scaling factors to preserve the physics information in the structure of the matrix. In my initial concept of automatic scaling I thought we would scale every matrix row individually such that each row would have a maximum absolute value of unity, but @jedbrown pointed out that this would render preconditioners like multigrid unusable. So now we scale by variable. Perhaps extending it to group similar variables also helps the preconditioner perform optimally.

jedbrown replied:
One can recover, but you'd need `MatSetNearNullSpace` so AMG knows what it needs to approximate on the coarse grid. If you use preconditioned residual norm, these scaling problems roughly go away (modulo numerical stability).

bwspenc commented on modules/contact/doc/content/modules/contact/index.md:
```diff
@@ -270,3 +270,38 @@ The recommended PETSc options for use with `NodeFaceConstraint` based contact ar
+
+- The solve efficiency of kinematic and tangential penalty formulations is independent of scaling (within the window of this problem)
+- Because RANFS constraint residuals/Jacobians are of gap magnitude, the RANFS solve really does perform better when the internal physics is scaled to be on the same order of magnitude
+- There are some bugs with the RANFS Jacobian functions because RANFS takes more non-linears and linears than kinematic, but with a FD Jacobian RANFS takes significantly less non-linears and linears
```
This needs to be tightened up a bit.  What do you mean by 'the window if this problem'? Change 'linears' and 'non-linears' to 'linear and nonlinear iterations'.

lindsayad replied:
> What do you mean by 'the window if this problem'?

I don't know what I mean. The whole idea behind automatic scaling was to improve the condition number of the operator and hence to improve the convergence of the linear solve (and hence to improve the convergence of the nonlinear solve). But recent weeks have shown that simply improving the condition number doesn't necessitate an improvement in the linear solve. That's what we're seeing here for kinematic and for tangential penalty and to a much more exaggerated extent in https://github.com/idaholab/moose/issues/12601#issuecomment-580574398. Though the condition number changes with scaling, the linear solve proceeds identically. So through discussions with @fdkong I've come to understand that the spectrum of eigenvalues, and how well they are clustered, is a better predictor of linear solve efficiency than just a single condition number. However, it seems that even the relationship between the eigenvalue spectrum and convergence of Krylov methods is still an open area of research.

bwspenc reviewed (changes requested):
You got me with the teardrop emoticon.

So it looks like the residual calculation was actually OK, but the Jacobian was wrong?  Does this mean that we still use the same penalty factor regardless of whether scaling is used?  If so, that's really good!

You might want to consider squashing some commits -- that's a pretty messy history.

lindsayad commented:
> So it looks like the residual calculation was actually OK, but the Jacobian was wrong?

Correct!

>  Does this mean that we still use the same penalty factor regardless of whether scaling is used? If so, that's really good!

Correct!

lindsayad committed:
Manual handling of Jacobians in NonlinearSystemBase::constraintJacobians

lindsayad committed:
Allow querying of displaced problem/systems for whether scaling is being performed

lindsayad committed:
Add ability to group variables for scaling

lindsayad committed:
change placement of scaling calculation

Move to NonlinearSystem::solve after the setting of an
initial guess based on preset BCs and predictors. This is
very beneficial when doing residual based scaling because
when there is loading from DirichletBCs the initial solution
after presetting will lead to non-zero residual entries from
volumetric objects.

This change *may* have negative side effects because I feel like
I moved the scaling computation to timestepSetup for a reason...

lindsayad committed:
Changes in RANFSNormalMechanicalContact

- By default, don't institute ping-ponging protection for RANFS
- Allow normal smoothing in RANFS contact

lindsayad committed:
Make kinematic Jacobians correct

Closes #14546

lindsayad committed:
Scaling tests

- RANFS
- Kinematic
- Tangential penalty

lindsayad committed:
Add ping-ponging tests

This batch of tests illustrates the susceptiblity of pure
node-face mechanical contact algorithms to ping-ponging. In this
case the ping-ponging comes from oscillation of a slave node
between two different master faces. A hybrid scheme in which
LM residuals are computed at slave nodes while primal residuals
are computed on mortar integrals does not display oscillations.
This does not necessarily mean that the hybrid scheme is immune
to oscillations, but it at least performs better on this test
case.

Refs #10950

lindsayad committed:
Document impact of scaling on convergence

lindsayad commented:
> I just merged a PR in today that makes contact depend on tensor_mechanics. It would make me feel so much better if these all used tensor_mechanics.

I updated all tests to use tensor mechanics.

lindsayad commented:
This is some cool stuff @fdkong. Posting more scaling findings when running the kinematic test case input file created in this PR. Summary for PJFNK:

- with preset boundary conditions, the nonlinear solve history is identical whether you are scaling or not scaling
- when boundary conditions are not preset, the nonlinear solve is ***much*** better when using  scaling as compared to when not using scaling. Moreover, the nonlinear solve is better when not using preset bcs and when using scaling, than when using preset bcs. When not using scaling, not using preset bcs is worse than when using preset bcs.

| Test case | total linears | total nonlinears |
| --- | --- | --- |
| preset, no scaling | 258 | 58 |
| preset, with scaling | 258 | 58 |
| not preset, no scaling | 310 | 61 |
| not preset, with scaling | 241 | 54 |

[with-scaling-preset-pjfnk.log](https://github.com/idaholab/moose/files/4172248/with-scaling-preset-pjfnk.log)
[no-scaling-preset-pjfnk.log](https://github.com/idaholab/moose/files/4172249/no-scaling-preset-pjfnk.log)
[no-scaling-not-preset-pjfnk.log](https://github.com/idaholab/moose/files/4172250/no-scaling-not-preset-pjfnk.log)
[with-scaling-not-preset-pjfnk.log](https://github.com/idaholab/moose/files/4172251/with-scaling-not-preset-pjfnk.log)

lindsayad commented:
Now with NEWTON. Same conclusions as for PJFNK.

| Test case | Total linears | total nonlinears |
| --- | --- | --- |
| preset, no scaling | 458 | 114 |
| preset, with scaling | 458 | 114 |
| not preset, no scaling | 625 | 145 |
| not preset, with scaling | 428 | 106 |

[with-scaling-not-preset-newton.log](https://github.com/idaholab/moose/files/4172864/with-scaling-not-preset-newton.log)
[no-scaling-not-preset-newton.log](https://github.com/idaholab/moose/files/4172865/no-scaling-not-preset-newton.log)
[no-scaling-preset-newton.log](https://github.com/idaholab/moose/files/4172867/no-scaling-preset-newton.log)
[with-scaling-preset-newton.log](https://github.com/idaholab/moose/files/4172868/with-scaling-preset-newton.log)

lindsayad commented:
It would be really nice to get this merged because the `scaling_group_variables` is very generic and I'd like to use it in more general (variable) scaling studies
